File: Remote/Source/Bank/BankObject/BankObject Subclasses/Manufacturer.m
//
// Manufacturer.m
// Remote
//
// Created by Jason Cardwell on 7/7/12.
// Copyright (c) 2012 Moondeer Studios. All rights reserved.
//

#import "Manufacturer.h"
#import "BankObjectGroup.h"

@implementation Manufacturer

@dynamic codesets, name;

+ (BOOL)isEditable { return NO; }

+ (BOOL)isPreviewable { return NO;}

+ (NSString *)directoryLabel { return @"Manufacturer"; }

+ (NSOrderedSet *)directoryItems { return nil; }

- (UIImage *)thumbnail { return nil; }

- (UIImage *)preview { return nil; }

- (NSString *)category { return nil; }

- (UIViewController *)editingViewController { return nil; }

- (NSOrderedSet *)subBankables { return nil; }

+ (instancetype)fetchManufacturerWithName:(NSString *)name context:(NSManagedObjectContext *)context
{
    __block Manufacturer * manufacturer = nil;

    [context performBlockAndWait:^{
        NSFetchRequest * fetchRequest = NSFetchRequestFromClassWithPredicate(@"name == %@", name);
        NSArray * fetchedObjects = [context executeFetchRequest:fetchRequest error:nil];
        manufacturer = [fetchedObjects lastObject];
    }];
    return manufacturer;
}

+ (instancetype)manufacturerWithName:(NSString *)name context:(NSManagedObjectContext *)context
{
    assert(name && context);

    __block Manufacturer * manufacturer = [self fetchManufacturerWithName:name context:context];

    if (!manufacturer) {
        [context performBlockAndWait:
         ^{
             manufacturer = [self MR_createInContext:context];
             manufacturer.name = name;
         }];
    }
    
    return manufacturer;
}

@end

File: RemoteData/Bank/Bankables/Manufacturer.swift
//
//  Manufacturer.swift
//  Remote
//
//  Created by Jason Cardwell on 9/30/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(Manufacturer)
class Manufacturer: IndexedBankCategoryObject, Detailable {


  @NSManaged var codeSets: Set<IRCodeSet>
  @NSManaged var devices: Set<ComponentDevice>

  var indexedSubcategories: [IndexedBankCategory] { return Array(codeSets) }
  func setIndexedSubcategories(subcategories: [IndexedBankCategory]) {
    if let codeSets = subcategories as? [IRCodeSet] { self.codeSets = Set(codeSets) }
  }

  /**
  manufacturerWithName:context:

  :param: name String
  :param: context NSManagedObjectContext

  :returns: Manufacturer
  */
  class func manufacturerWithName(name: String, context: NSManagedObjectContext) -> Manufacturer {
    var manufacturer: Manufacturer!
    context.performBlockAndWait { () -> Void in
      manufacturer = self.objectWithValue(name, forAttribute: "name", context: context)
      if manufacturer == nil {
        manufacturer = self.createInContext(context)
        manufacturer.name = name
      }
    }
    return manufacturer
  }

  /**
  updateWithData:

  :param: data [String AnyObject]
  */
  override func updateWithData(data: [String:AnyObject]) {
    super.updateWithData(data)
    updateRelationshipFromData(data, forKey: "codeSets", lookupKey: "code-sets")
    updateRelationshipFromData(data, forKey: "devices")
  }

  /**
  rootCategoryNamed:context:

  :param: name String
  :param: context NSManagedObjectContext

  :returns: IndexedBankCategory?
  */
  override class func rootCategoryNamed(name: String, context: NSManagedObjectContext) -> IndexedBankCategory? {
    return objectWithValue(name, forAttribute: "name", context: context)
  }

  class var rootCategory: BankRootCategory<BankCategory,BankModel> {
    let manufacturers = objectsInContext(DataManager.rootContext, sortBy: "name") as? [Manufacturer]
    return BankRootCategory(label: "Manufacturers",
                             icon: UIImage(named: "1022-factory")!,
                             items: manufacturers ?? [],
                             editableItems: true)
  }

  /**
  detailController

  :returns: UIViewController
  */
  func detailController() -> UIViewController { return ManufacturerDetailController(model: self) }

}

extension Manufacturer: MSJSONExport {

  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()

    appendValueForKeyPath("devices.commentedUUID",   forKey: "devices", toDictionary: dictionary)
    appendValueForKeyPath("codeSets.JSONDictionary", forKey: "code-sets", toDictionary: dictionary)

    dictionary.compact()
    dictionary.compress()

    return dictionary
  }

}
